### PR TITLE
[pytorch] Add private `dispatchless_custom_op`

### PR DESCRIPTION
**Pull request by njriasan**
Summary:
Implement the functional, private-API scope of https://github.com/pytorch/pytorch/issues/193056 in fbcode PyTorch. Eager calls invoke the backing Python function without custom-op dispatch or pytree flattening. `make_fx` and Dynamo preserve an opaque `flat_dispatchless_call` with per-call pytree specifications. Inductor decomposes the call into its backing implementation, enabling fusion and ordinary autograd.

The direct-call higher-order operator reuses existing pytree/flat-apply helpers without changing dispatcher registration or the global decomposition table. Functions must be functional, FakeTensor-compatible, and receive tensor dependencies explicitly. Mutation, output aliases, custom gradient registration, and export serialization are out of scope; other AOT backends must provide decomposition before differentiation.

Registered frozen-dataclass constants exposed an existing Dynamo gap: `FrozenDataClassVariable.as_python_constant` bypassed the base class's registered-constant path. Preserve that path and its existing equality guards without allowing arbitrary frozen-dataclass reconstruction. Regression coverage checks changed constants recompile and equal constants reuse a graph.

Authored with assistance from OpenAI Codex. Submitted as a draft for human review.

Test Plan:
36 tests passed with no failures or skips, covering eager execution, tracing modes, pytree inputs/outputs, guarding, dynamic shapes, CPU/CUDA values and gradients, Inductor decomposition/fusion, and existing `flat_apply` regressions:
```
buck2 test mode/opt fbcode//caffe2/test/dynamo:test_dynamo fbcode//caffe2/test/inductor:dispatchless fbcode//caffe2/test/inductor:dispatchless_cpu -- --regex 'test_(dispatchless|flat_apply)'
```
https://www.internalfb.com/intern/testinfra/testrun/31525197424747472

9 additional frozen-dataclass and registered-constant regression tests passed with no failures or skips:
```
buck2 test mode/opt fbcode//caffe2/test/dynamo:test_dynamo -- test_proxy_frozen_dataclass test_frozen_dataclass_default_value test_pytree_register_constant_with_side_effect test_frozen_dataclass_as_dict_key
```
https://www.internalfb.com/intern/testinfra/testrun/15199648939438376

Lint passed:
```
arc lint -a --only-changed caffe2/torch/_library/dispatchless.py caffe2/torch/_dynamo/variables/torch.py caffe2/torch/_dynamo/variables/builder.py caffe2/torch/_dynamo/variables/user_defined.py caffe2/torch/_inductor/decomposition.py caffe2/test/dynamo/test_dispatchless.py caffe2/test/inductor/test_dispatchless.py caffe2/test/inductor/BUCK
```
`arc f` and the extra `CITRINEAGENT` lint invocation reported no configured linters. `arc pyre check-owning-targets` found six owning targets, but type checking is disabled for all six, so no type checks ran. Tests use `mode/opt`; the default dev mode encountered unrelated Dynamo collection errors and an ASAN/CUDA initialization incompatibility.

Benchmark-only follow-up (2026-09-09): implementation and functional test sources are unchanged; the earlier 45 passing tests remain the functional validation evidence. Added a runnable benchmark in `caffe2/benchmarks/dynamo/microbenchmarks/dispatchless_custom_op.py`.

```
buck2 run mode/opt fbcode//caffe2/benchmarks/dynamo/microbenchmarks:dispatchless_custom_op -- --min-run-time 1 --repeats 3
```
Buck run: https://www.internalfb.com/buck2/3a763aeb-8475-4a47-a6c6-3e06f1765872

Exit 0. Output equality checks passed in all 27 variant/case combinations; gradient checks passed in all requires-grad cases. All three variants execute the same non-aliasing `x.clone()` body. The standard `custom_op` has the matching clone backward registered. This measures warmed-up CPU forward calls, with registration, explicit warmup, correctness checks, and backward excluded. The requires-grad case includes forward autograd graph construction. There are no timing assertions.

Environment: AMD EPYC 9654 96-Core Processor, host `devgpu007.eag6.facebook.com`, PyTorch `2.15.0a0+fb`, Python `3.12.14+meta`, CPU float32, one intra-op thread, `AVX2` capability. `Timer.blocked_autorange` ran at least 1 second per variant per round, for 3 rounds with rotated variant order. The CPU timer does not synchronize accelerators.

All latency columns are microseconds per call. Cells show median (IQR), pooled over per-call block timings. Delta columns subtract the corresponding plain-function median; speedup is total custom-op latency divided by dispatchless latency.

| Mode | Elements | Plain us (IQR) | custom_op us (IQR) | Dispatchless us (IQR) | custom_op delta us | Dispatchless delta us | Speedup |
| --- | ---: | ---: | ---: | ---: | ---: | ---: | ---: |
| no_grad | 0 | 0.748 (0.020) | 6.955 (0.071) | 2.285 (0.012) | 6.208 | 1.537 | 3.04x |
| no_grad | 1 | 0.893 (0.004) | 7.125 (0.026) | 2.463 (0.016) | 6.233 | 1.570 | 2.89x |
| no_grad | 1024 | 0.945 (0.005) | 7.224 (0.038) | 2.588 (0.032) | 6.279 | 1.643 | 2.79x |
| grad_enabled | 0 | 0.749 (0.003) | 7.077 (0.044) | 2.278 (0.011) | 6.328 | 1.528 | 3.11x |
| grad_enabled | 1 | 0.912 (0.005) | 7.468 (0.092) | 2.517 (0.031) | 6.556 | 1.605 | 2.97x |
| grad_enabled | 1024 | 0.968 (0.005) | 7.489 (0.119) | 2.642 (0.028) | 6.520 | 1.674 | 2.83x |
| requires_grad | 0 | 1.020 (0.015) | 9.793 (0.052) | 2.623 (0.051) | 8.773 | 1.602 | 3.73x |
| requires_grad | 1 | 1.169 (0.012) | 9.921 (0.072) | 2.775 (0.019) | 8.752 | 1.606 | 3.58x |
| requires_grad | 1024 | 1.224 (0.025) | 10.092 (0.102) | 2.906 (0.032) | 8.868 | 1.681 | 3.47x |

`no_grad` disables grad mode; `grad_enabled` enables grad mode but inputs do not require gradients; `requires_grad` enables grad mode and input gradients.

Observed total-call speedup was 2.79-3.73x. Dispatchless still adds approximately 1.53-1.68 us above plain Python, compared with 6.21-8.87 us for `custom_op`. These are shared-devserver CPU microbenchmark measurements, not GPU, compilation, backward, or end-to-end workload speedups; IQRs describe variability, not confidence intervals.

Follow-up lint:
```
arc lint -a --only-changed caffe2/benchmarks/dynamo/microbenchmarks/dispatchless_custom_op.py caffe2/benchmarks/dynamo/microbenchmarks/BUCK
arc lint -a --rev '.^' --only-changed
```
No lint issues. `arc f` and `CITRINEAGENT` reported no configured linters. `arc pyre check-owning-targets --reuse-current-config --report-unchecked-targets caffe2/benchmarks/dynamo/microbenchmarks/dispatchless_custom_op.py` found two owning targets with typing disabled, so no type checks ran.

Differential Revision: D119414357

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo @azahed98

### CODE REVIEW

**Comment by njriasan** [comment]
I haven't evaluated the code yet for quality. I'm just trying to get the OSS CI running first.
